explainer-datamodel.md

```diff
@@ -0,0 +1,413 @@
+
+On the web today, if you want to render text, you essentially have two options (generally): use
+the DOM (HTML's CharacterData/Text nodes or SVG Text-related Elements), in which case you put the
+text into the "retained" mode infrastructure of the web platform and it decides when and how
+to compose and render the text with declarative input from you (in the form of CSS); or you can
+use Canvas and "write" the text when and how you want with JavaScript (an "immediate" mode 
+approach). The Canvas provides very limited text support today and (by design) leaves any special
+formatting, text wrapping, international support, etc., up to the JavaScript author.
+
+We propose a hybrid approach that allows the author to retain the benefits of the DOM's "retained"
+mode objects, reusing the web platform's existing line formatting and wrapping engine along with 
+support for CSS, while also enabling authors to apply the objects to a Canvas how and when they
+choose. Alternately, authors may not desire to render the objects to a Canvas at all, and instead
+can run a given layout over them to obtain metrics useful in other scenarios.
+
+In this explainer, a new formatted text data model is presented (the "retained" mode objects) which
+is the minimal necessary objects to allow CSS-styled text runs to be collected in a sequence
+for layout processing.
+
+### Principles
+* An imperative JavaScript-friendly text representation.
+* Scope to the needs of inline text layout.
+* Leverage CSS as the universal text layout system for the web (even in an imperative model) which also provides for future extensibility.
+* Avoid multi-level hierarchal text structures (which must be linearly iterated for rendering anyway)
+* Object-centric model (versus range+indexes) to improve encapsulation and avoid problems with overlapping formatting.
+
+### Primary Objects
+
+The data model for formatted text is quite simple and involves only two objects: `FormattedText`
+is a container that orders a list of `FormattedTextRun` objects that contain the actual text and
+formatting information. (The `FormattedText` container can also have formatting applied that may
+inherit down to the `FormattedTextRun`s.)
+
+To maximize authoring convenience, we are proposing using the 
+[ObserverableArray pattern](https://heycam.github.io/webidl/#idl-observable-array) to allow authors
+to manage the contained array exactly like a JavaScript Array.
+
+<img src="explainerresources/formatted-text-om.png" alt="FormattedText object holds an array of FormattedTextRun objects through a property called 'textruns'." align="center"/>
+
+## Creating FormattedText and FormattedTextRun objects
+
+`FormattedText` and `FormattedTextRun` objects are created using the 'new' pattern:
+
+```js
+let text = new FormattedText();
+let textrun = new FormattedTextRun();
+```
+
+Attach `FormattedTextRun` objects to their `FormattedText` object's `textruns` property (an Array):
+
+```js
+text.textruns.push( textrun );
```

Comment:
FYI, this is currently exposed as text.appendRun(), text.deletRun(). Indexing operators work like text[index] and styleMap is exposed as text.styleMap.

   var text2 = new CanvasFormattedText();
   text2.appendRun(runB);
   assert_exception(()=> { text2[0] = text[0]; 

Ill take an action item to fix the implementation.

Reply:
I'm going to start a separate MD page with our implementation notes/gaps for public reference. I'd like to have a reference for folks that want to know how to play with the feature immediately while we align the spec/implementation :).

Reply:
See https://github.com/WICG/canvas-formatted-text/commit/742d9b4c4c1f454ffe47383aafb803a169bdae98